src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,388 @@
+          return kj::READY_NOW;
+        }
+
+        // We will try to get the function, if we can't we'll throw an error to the client.
+        auto maybeFn = tryGetFn(lock, entrypointName, ctx, methodName);
+        KJ_IF_SOME(e, maybeFn.tryGet<kj::String>()) {
+          auto serialized = serializeV8(js, js.error(e));
+          auto builder = callContext.initResults(
+              capnp::MessageSize { serialized.size() / 8 + 8, 0 }).initException();
+          builder.setV8Serialized(kj::mv(serialized));
+          return kj::READY_NOW;
         }
-    ));
+
+        auto fn = KJ_REQUIRE_NONNULL(maybeFn.tryGet<v8::Local<v8::Function>>());
+
+        // We have a function, so let's call it and serialize the result for RPC.
+        // We account for both the JS method throwing and our serialization attempt failing.
+        // If the function returns a promise we will wait for the promise to finish so we can
+        // serialize the result.
+        // TODO(now): Do we need to use awaitJs here? `call()` has to return a kj::Promise, so maybe
+        // we can return some other promise that resolves when this jsg::Promise does?
+        return ctx.awaitJs(js, callFuncAndSerialize(js, fn, serializedArgs).then(js,
+            [&](jsg::Lock& js, SerializedResult result) {
+          KJ_SWITCH_ONEOF(result) {
+            KJ_CASE_ONEOF(ok, Returned) {
+              auto builder = callContext.initResults(capnp::MessageSize {
+                  ok.serializedValue.size() / 8 + 8, 0 });
+              builder.initValue().setV8Serialized(kj::mv(ok.serializedValue));
+            }
+            KJ_CASE_ONEOF(err, ThrewException) {
+              auto builder = callContext.initResults(capnp::MessageSize {
+                  err.serializedException.size() / 8 + 8, 0 });
+              builder.initException().setV8Serialized(kj::mv(err.serializedException));
+            }
+          }
+        }));
+      }).then([this]() {
+        // We want to fulfill the callPromise so customEvent can continue executing.
+        callFulfiller->fulfill();
+      });
+    } catch(...) {
+      auto e = kj::getCaughtExceptionAsKj();
+      if (auto desc = e.getDescription();
+          !jsg::isTunneledException(desc) && !jsg::isDoNotLogException(desc)) {
+        LOG_EXCEPTION("JsRpcTargetCall"_kj, e);
+      // We want to fulfill the callPromise so customEvent can continue executing.
+      callFulfiller->fulfill();
+      }
+    }
+    return kj::READY_NOW;
   }
+
+  KJ_DISALLOW_COPY_AND_MOVE(JsRpcTargetImpl);
+
+private:
+  // Prevents us from calling a function defined on the Object Prototype, or a registered method.
+  bool verifyCallable(jsg::Lock& js, v8::Local<v8::Object> workerObj, kj::StringPtr name) {
+    // Walk down the prototype chain until we get the Object prototype.
+    auto currObj = workerObj->GetPrototype().As<v8::Object>();
+    while (currObj->GetPrototype()->IsObject()) {
+      currObj = currObj->GetPrototype().As<v8::Object>();
```

Comment:
This is not guaranteed to give you to the intrinsic `Object` prototype unfortunately. It could be that the object is created using `Object.create(null)` for instance. I think what you actually want here is to create a new `v8::Object` (you can use `v8::Local<v8::Object> obj = js.obj()`) and grab its prototype (which is guaranteed to be the intrinsic `Object` prototype)